wip-trampoline.md

```diff
@@ -0,0 +1,780 @@
+trampoline nodes is deferred to the trampoline nodes themselves.
+
+Merchants may include a few trampoline nodes that are close to them in their invoice to help payers select a good last
+trampoline node. This is optional and payers aren't required to use this routing hint.
+
+## Table of Contents
+
+* [Features](#features)
+* [Packet Structure](#packet-structure)
+  * [Trampoline Onion](#trampoline-onion)
+  * [Paying via trampoline nodes](#paying-via-trampoline-nodes)
+  * [Paying to non-trampoline nodes](#paying-to-non-trampoline-nodes)
+* [Routing Gossip](#routing-gossip)
+  * [Fees and CLTV estimation](#fees-and-cltv-estimation)
+  * [The `node_update` Message](#the-node_update-message)
+  * [Filtering gossip messages](#filtering-gossip-messages)
+* [Fees and CLTV requirements](#fees-and-cltv-requirements)
+* [Appendix A: Examples](#appendix-a-examples)
+  * [Merchant supporting trampoline payments](#merchant-supporting-trampoline-payments)
+  * [Merchant without trampoline support](#merchant-without-trampoline-support)
+* [Appendix B: Trampoline and Rendez-Vous](#appendix-b-trampoline-and-rendez-vous)
+* [Appendix C: Trampoline and MPP](#appendix-c-trampoline-and-mpp)
+* [References](#references)
+
+## Features
+
+Trampoline routing uses the following `globalfeatures` flags:
+
+| Bits  | Name                 | Description                                    | Link |
+|-------|----------------------|------------------------------------------------|------|
+| 10/11 | `trampoline_routing` | This node requires/supports trampoline routing | TBD  |
+
+Feature flags are assigned in pairs, but the mandatory value may not be applicable here.
+
+Gossip messages filtering uses the following `localfeatures` flags:
+
+| Bits  | Name                 | Description                                        | Link |
+|-------|----------------------|----------------------------------------------------|------|
+| 12/13 | `gossip_filters`     | This node requires/supports gossip message filters | TBD  |
+
+## Packet Structure
+
+### Trampoline Onion
+
+The trampoline onion is a fixed-size tlv packet with the following structure:
+
+1. type: 14 (`trampoline_onion_packet`)
+2. data:
+   * [`byte`:`version`]
+   * [`point`:`public_key`]
+   * [`400*byte`:`hop_payloads`]
+   * [`32*byte`:`hmac`]
+
+This results in a `466`-bytes fixed-size tlv where the value has exactly the same format as the `onion_packet` with
+a smaller `hop_payloads` (`400` bytes instead of `1300`).
+
+Trampoline `hop_payload`s may contain the following fields:
+
+1. tlvs: `trampoline_tlv_payload`
+2. types:
+    * type: 2 (`amt_to_forward`)
+    * data:
+        * [`tu64`:`amt_to_forward`]
+    * type: 4 (`outgoing_cltv_value`)
+    * data:
+        * [`tu32`:`outgoing_cltv_value`]
+    * type: 10 (`next_node_id`)
+    * data:
+        * [`point`:`next_node_id`]
+    * type: 12 (`recipient_features`)
+    * data:
+        * [`length*byte`:`recipient_features`]
```

Comment:
Global and local are being merged into a single set of feature bits, which aligns well with that field.
I don't think invoice features are useful there; but maybe I'm missing a use-case (if that's the case, another tlv for invoice features can be added later).

Reply:
Cool for merging both local/global features. Yeah would be easy to upgrade but doesn't the last version of base AMP use invoice feature bits to signal support of feature?

Reply:
I don't think this will be an issue, it's an implementation detail. If the invoice specified that AMP could be used, the sender will include the `total_amount` TLV. That lets the trampoline node detect that AMP may be used to route to the destination (without the need for an explicit feature bit). But I agree that some future invoice feature bits might require us to add an `invoice_features` TLV here.

Reply:
This isn't a concern anymore as invoice features now use the same bit namespace as other features. Yay!